lib/oauthdb.js

```diff
@@ -0,0 +1,177 @@
+ *
+ * Currently this is not actually talking to a database,
+ * it's making authenticated calls to the fxa-oauth-server API
+ * to interrogate and manipulate its state, essentially treating
+ * fxa-oauth-server as a kind of backend micro-service.
+ *
+ * We want to work towards merging the fxa-oauth-server code
+ * directly into the main fxa-auth-server process, at which point
+ * this abstraction will convert into more direct db access.
+ *
+ */
+
+const Joi = require('joi')
+const JWT = require('jsonwebtoken')
+
+const P = require('./promise')
+const createBackendServiceAPI = require('./backendService')
+const error = require('./error')
+const validators = require('./routes/validators')
+
+// The oauth-server's error numbers overlap and conflict
+// with the auth-server's, so we have to map them to new ones.
+function mapOAuthError(err) {
+  if (err instanceof error) {
+    return err
+  }
+  switch (err.errno) {
+    case 101:
+      return error.unknownClientId(err.clientId)
+    case 119:
+      return error.staleAuthAt(err.authAt)
+    default:
+      return error.unexpectedError()
+  }
+}
+
+module.exports = (log, config) => {
+
+  const OAuthAPI = createBackendServiceAPI(log, config, 'oauth', {
+
+    getClientInfo: {
+      path: '/v1/client/:clientId',
+      method: 'GET',
+      validate: {
+        params: {
+          clientId: validators.clientId.required()
+        },
+        response: {
+          id: validators.clientId.required(),
+          name: Joi.string().max(25).regex(validators.DISPLAY_SAFE_UNICODE).required(),
+          trusted: Joi.boolean().required(),
+          image_uri: Joi.string().optional().allow(''),
+          redirect_uri: Joi.string().required().allow('')
+        }
+      }
+    },
+
+    getScopedKeyData: {
+      path: '/v1/key-data',
+      method: 'POST',
+      validate: {
+        payload: {
+          client_id: validators.clientId.required(),
+          assertion: validators.assertion.required(),
+          scope: validators.scope.required()
+        },
+        response: Joi.object().pattern(/^/, [
+          Joi.object({
+            identifier: validators.scope.required(),
+            keyRotationSecret: validators.hexString.length(64).required(),
+            keyRotationTimestamp: Joi.number().required(),
+          })
+        ])
+      }
+    }
+  })
+
+  // Make a symmetrically-signed JWT assertion that we can pass to
+  // fxa-oauth-server in lieu of a full-blown BrowserID assertion.
+
+  function makeAssertionJWT(credentials) {
+    if (! credentials.emailVerified) {
+      throw error.unverifiedAccount()
+    }
+    if (credentials.mustVerify && ! credentials.tokenVerified) {
+      throw error.unverifiedSession()
+    }
+    const opts = {
+      algorithm: 'HS256',
+      expiresIn: 60,
+      audience: config.oauth.url,
+      issuer: config.domain
+    }
+    const claims = {
+      uid: credentials.uid,
+      'fxa-generation': credentials.verifierSetAt,
```

Comment:
I have to ask, why does `fxa-generation` have so many names? I see [`fxa-generation`, `generation`](https://github.com/mozilla/fxa-auth-server/blob/d590d6e5f7f63f67226fe22abaedefaff75e745b/lib/signer.js#L24), and [`verifierSetAt`](https://github.com/mozilla/fxa-auth-server/blob/d590d6e5f7f63f67226fe22abaedefaff75e745b/lib/routes/account.js#L1036).

Reply:
I think reliers are supposed to treat the "generation number" as having extremely minimal semantics, along the lines of "if you see any auth tokens with a generation number below the max you've ever seen, they're invalid".  We don't necessarily want to promise that it always corresponds to `verifierSetAt`, although it does in practice right now.

(But I also believe we're not exposing this to OAuth reliers at all, so it's a bit of a moot point going forward).